### PR TITLE
Added explanation for E0391

### PR DESCRIPTION
**Pull request by soon**
Part of #24407

### CODE REVIEW

File: src/librustc_typeck/diagnostics.rs [hunk 1]
```diff
@@ -2215,6 +2215,23 @@ For more information see the [opt-in builtin traits RFC](https://github.com/rust
 -lang/rfcs/blob/master/text/0019-opt-in-builtin-traits.md).
 "##,
 
+E0391: r##"
+This error indicates that some types or traits depends on each other
+and therefore could not be constructed properly.
+
+The following example contains circular dependency between two traits:
```
Comment (nrc) [inline src/librustc_typeck/diagnostics.rs]: nit: "contains a circular"

File: src/librustc_typeck/diagnostics.rs [hunk 2]
```diff
@@ -2215,6 +2215,23 @@ For more information see the [opt-in builtin traits RFC](https://github.com/rust
 -lang/rfcs/blob/master/text/0019-opt-in-builtin-traits.md).
 "##,
 
+E0391: r##"
+This error indicates that some types or traits depends on each other
+and therefore could not be constructed properly.
```
Comment (nrc) [inline src/librustc_typeck/diagnostics.rs]: and another couple of nits: s/could not/cannot, s/ properly//

File: src/librustc_typeck/diagnostics.rs [hunk 3]
```diff
@@ -2215,6 +2215,23 @@ For more information see the [opt-in builtin traits RFC](https://github.com/rust
 -lang/rfcs/blob/master/text/0019-opt-in-builtin-traits.md).
 "##,
 
+E0391: r##"
+This error indicates that some types or traits depends on each other
```
Comment (nrc) [inline src/librustc_typeck/diagnostics.rs]: depend

**Comment by nrc** [comment]
Thanks for the PR, looks like an improvement. I suggested a few minor grammar improvements. I wonder if you could find an example that has a straightforward solution? That is often informative for readers when they get an error like this. (I'm not exactly sure when this error gets triggers, otherwise I'd suggest something myself).

**Comment by soon** [comment]
@nrc There is an [issue](https://github.com/rust-lang/rust/issues/23118) related to the same error. What do you think, are code samples from that issue easy enough to be presented in the documentation (maybe, with small fixes)?

**Comment by nrc** [comment]
Hmm, I don't think that is too useful. In fact, I don't think there is going to be a really informative example here. Looks good as is.

**Pull request merged**